File: lib/domain/usecases/branch/merge_branch_usecase.dart
import '../../entities/enums.dart';
import '../../entities/commit.dart' as entity;
import '../../repositories/i_branch_repository.dart';
import '../../repositories/i_commit_repository.dart';
import '../../repositories/i_task_repository.dart';

/// 合并分支用例
class MergeBranchUseCase {
  MergeBranchUseCase(
    this._branchRepository,
    this._taskRepository,
    this._commitRepository,
  );

  final IBranchRepository _branchRepository;
  final ITaskRepository _taskRepository;
  final ICommitRepository _commitRepository;

  Future<void> execute({
    required String sourceBranchId,
    required String targetBranchId,
  }) async {
    if (sourceBranchId == targetBranchId) {
      throw ArgumentError('不能合并分支到自身');
    }

    // 1. 验证分支存在
    final sourceBranch =
        await _branchRepository.getById(sourceBranchId);
    if (sourceBranch == null) {
      throw ArgumentError('源分支不存在: $sourceBranchId');
    }

    final targetBranch =
        await _branchRepository.getById(targetBranchId);
    if (targetBranch == null) {
      throw ArgumentError('目标分支不存在: $targetBranchId');
    }

    // 2. 获取源分支的所有任务
    final tasks =
        await _taskRepository.getByBranchId(sourceBranchId);

    // 3. 将未完成的任务移动到目标分支
    for (final task in tasks) {
      if (task.status != TaskStatus.done) {
        await _taskRepository.update(
          task.copyWith(branchId: targetBranchId),
        );

        // 记录合并提交
        await _commitRepository.create(
          entity.Commit.create(
            taskId: task.id,
            branchId: targetBranchId,
            message: '合并自分支: ${sourceBranch.name}',
            type: CommitType.merge,
          ),
        );
      }
    }

    // 4. 删除源分支（软删除）
    await _branchRepository.delete(sourceBranchId);
  }
}

File: web/src/application/usecases/branch/merge-branch-usecase.ts
import { Commit } from '../../../domain/entities/commit';
import { CommitType } from '../../../domain/entities/enums';
import { IBranchRepository } from '../../../domain/repositories/i-branch-repository';
import { ICommitRepository } from '../../../domain/repositories/i-commit-repository';
import { ITaskRepository } from '../../../domain/repositories/i-task-repository';

export class MergeBranchUseCase {
  constructor(
    private branchRepo: IBranchRepository,
    private taskRepo: ITaskRepository,
    private commitRepo: ICommitRepository
  ) {}

  async execute(sourceId: string, targetId: string): Promise<void> {
    const source = await this.branchRepo.getById(sourceId);
    const target = await this.branchRepo.getById(targetId);
    if (!source || !target) {
      throw new Error('分支不存在');
    }
    if (source.repositoryId !== target.repositoryId) {
      throw new Error('只能合并同一仓库的分支');
    }

    const tasks = await this.taskRepo.getByBranchId(sourceId);
    const incompleteTasks = tasks.filter((task) => !task.isCompleted);

    for (const task of incompleteTasks) {
      const moved = task.copyWith({ branchId: targetId });
      await this.taskRepo.update(moved);
    }

    const mergeCommit = Commit.create(
      tasks.length > 0 ? tasks[0].id : sourceId,
      targetId,
      `合并分支 ${source.name} 到 ${target.name}`,
      CommitType.merge
    );
    await this.commitRepo.create(mergeCommit);
  }
}
